### PR TITLE
Add varchar support for min/max_by

### PR DESCRIPTION
**Pull request by xiaoxmeng**
Add varchar support for min/max_by with unit tests. The followup is to add e2e test coverage.

### CODE REVIEW

File: velox/exec/GroupingSet.cpp [hunk 1]
```diff
@@ -174,26 +174,15 @@ void GroupingSet::addInputForActiveRows(
...
-    for (int32_t i = 0; i < hashers.size(); ++i) {
-      auto key = input->loadedChildAt(hashers[i]->channel());
-      if (mode != BaseHashTable::HashMode::kHash) {
-        if (!hashers[i]->computeValueIds(*key, activeRows_, lookup_->hashes)) {
-          rehash = true;
-        }
-      } else {
-        hashers[i]->hash(*key, activeRows_, i > 0, lookup_->hashes);
+  }
+  for (int32_t i = 0; i < hashers.size(); ++i) {
```
Comment (mbasmanova) [inline velox/exec/GroupingSet.cpp:178]: Why is this change? I didn't expect any changes to the hash aggregation operator itself, only to the min/max_by aggregate functions.

File: velox/exec/tests/utils/QueryAssertions.cpp [hunk 2]
```diff
@@ -190,7 +190,9 @@ velox::variant variantAt<TypeKind::DATE>(
 
 template <TypeKind kind>
 velox::variant variantAt(const ::duckdb::Value& value) {
-  using T = typename KindToFlatVector<kind>::WrapperType;
```
Comment (mbasmanova) [inline velox/exec/tests/utils/QueryAssertions.cpp:193]: Curious, which type required this change.

File: velox/exec/GroupingSet.cpp (hunk 1 above)
Comment (Yuhta) [reply velox/exec/GroupingSet.cpp:178]: This looks like just code cleanup

Comment (mbasmanova) [reply velox/exec/GroupingSet.cpp:178]: > This looks like just code cleanup

@xiaoxmeng Is this the case? If so, would you submit it in a separate PR?

File: velox/exec/tests/utils/QueryAssertions.cpp (hunk 2 above)
Comment (xiaoxmeng) [reply velox/exec/tests/utils/QueryAssertions.cpp:193]: The wrapped type of VARCHAR is StringView and duckdb's ::duckdb::Value::GetValue doesn't support that template type.

File: velox/exec/GroupingSet.cpp (hunk 1 above)
Comment (xiaoxmeng) [reply velox/exec/GroupingSet.cpp:178]: Yeah, it is just a code cleanup. Will put it in a separate PR.

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 3]
```diff
@@ -426,47 +436,1466 @@ class MinByAggregate : public MinMaxByAggregate<T, U> {
...
+
+  void destroy(folly::Range<char**> groups) override {
+    for (auto group : groups) {
+      valueAccumulator(group)->destroy(allocator_);
+    }
+  }
+
+ protected:
+  template <typename MayUpdate>
+  void addRawInput(
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:529]: I wonder if this is a copy-paste from the original class. If so, is there a way to re-use the logic by provide a different updateValues lambda?

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 4]
```diff
@@ -426,47 +436,1466 @@ class MinByAggregate : public MinMaxByAggregate<T, U> {
...
+      }
+      comparisonValue(group) = newComparisonValue;
+    }
+  }
+
+  inline int32_t valueAccumulatorOffset() const {
+    return Aggregate::offset_;
+  }
+
+  inline SingleValueAccumulator* valueAccumulator(char* group) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:776]: nit: a convention is to call this method 'value'

**Review by mbasmanova — COMMENTED** [review]
@xiaoxmeng The implementation looks good, but there is a lot of copy paste. It feels like it should be possible to reduce that using templates.

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 5]
```diff
@@ -303,13 +307,15 @@ class MinMaxByAggregate : public exec::Aggregate {
       char* group,
       T newValue,
       U newComparisonValue,
-      bool isNull,
+      bool isValueNull,
       MayUpdate mayUpdate) {
     clearNull(group);
     if (mayUpdate(comparisonValue(group), newComparisonValue)) {
-      value(group) = newValue;
+      valueIsNull(group) = isValueNull;
```
[the remark this answers is not in the document]
Comment (xiaoxmeng) [reply velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:314]: If isValueNull is null, we don't need to set the value as it is null.

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 6]
```diff
@@ -13,89 +13,214 @@
...
+template <typename T, typename TAccumulator>
+void extract(
+    TAccumulator* accumulator,
+    const VectorPtr& vector,
+    vector_size_t index,
+    T* rawValues) {
+  if constexpr (isNumeric<T>()) {
+    rawValues[index] = *accumulator;
+  } else {
+    VELOX_DCHECK(accumulator->hasValue());
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:43]: This check is redundant since SingleValueAccumulator::read already has it:

```
void SingleValueAccumulator::read(const VectorPtr& vector, vector_size_t index)
    const {
  VELOX_CHECK(begin_);
...
bool SingleValueAccumulator::hasValue() const {
  return begin_ != nullptr;
}
```

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 7]
```diff
@@ -13,89 +13,214 @@
...
+  if constexpr (isNumeric<T>()) {
+    *accumulator = decodedVector.valueAt<T>(index);
+  } else {
+    accumulator->write(
+        decodedVector.base(), decodedVector.index(index), allocator);
+  }
+}
+
+template <typename T, typename TAccumulator>
+bool greaterThan(
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:63]: Method-level comment and a comment inside the body of the function could be useful. It is confusing to the reader that one branch returns the result of > , but the other < 0.

```
/// Returns true if the value in 'index' row of 'newComparisons' is strictly greater than the value in the 'accumulator'.
...

// SingleValueAccumulator::compare has the semantics of accumulator-value-is-less-than-vector-value.
return !accumulator->hasValue() ||
        (accumulator->compare(newComparisons, index) < 0);
```

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 8]
```diff
@@ -13,89 +13,214 @@
...
+  if constexpr (isNumeric<T>()) {
+    return newComparisons.valueAt<T>(index) > *accumulator;
+  } else {
+    return !accumulator->hasValue() ||
+        (accumulator->compare(newComparisons, index) < 0);
+  }
+}
+
+template <typename T, typename TAccumulator>
+bool lessThan(
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:76]: Comments here would be helpful as well.

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 9]
```diff
@@ -13,89 +13,214 @@
...
 template <typename T, typename U>
 class MinMaxByAggregate : public exec::Aggregate {
  public:
+  using ValueAccumulatorType =
+      typename AccumulatorTypeTraits<T>::AccumulatorType;
+  using ComparisonAccumulatorType =
+      typename AccumulatorTypeTraits<U>::AccumulatorType;
+
+  /// NOTE: the passed min/max limit is only meaningful is comparison type U is
+  /// of numeric type.
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:116]: typo: of -> is a

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 10]
```diff
@@ -13,89 +13,214 @@
...
+    return sizeof(ValueAccumulatorType) + sizeof(ComparisonAccumulatorType) +
+        sizeof(bool);
   }
 
   void initializeNewGroups(
       char** groups,
       folly::Range<const vector_size_t*> indices) override {
     exec::Aggregate::setAllNulls(groups, indices);
-    for (auto i : indices) {
+    for (const int i : indices) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:129]: nit: const auto or const vector_size_t

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 11]
```diff
@@ -13,89 +13,214 @@
...
       auto group = groups[i];
-      comparisonValue(group) = initialValue_;
       valueIsNull(group) = true;
+
+      if (!isNumeric<T>()) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:133]: constexpr

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 12]
```diff
@@ -13,89 +13,214 @@
...
+        new (group + offset_) SingleValueAccumulator();
+      }
+
+      if constexpr (isNumeric<U>()) {
+        *comparisonValue(group) = initialValue_;
+      } else {
+        new (
+            group + offset_ +
+            (isNumeric<T>() ? sizeof(T) : sizeof(SingleValueAccumulator)))
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:142]: sizeof(ValueAccumulatorType)

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 13]
```diff
@@ -13,89 +13,214 @@
...
+    }
+    U* rawComparisonValues = nullptr;
+    if constexpr (isNumeric<U>()) {
+      auto flatComparisonVector = comparisonVector->as<FlatVector<U>>();
+      VELOX_CHECK(flatComparisonVector != nullptr);
+      rawComparisonValues = flatComparisonVector->mutableRawValues();
+    }
     BufferPtr nulls = valueVector->mutableNulls(rowVector->size());
-    uint64_t* nullValues = nulls->asMutable<uint64_t>();
+    uint64_t* rawValueNulls = nulls->asMutable<uint64_t>();
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:198]: nit: rawValueNulls -> rawNulls (to match 'nulls')

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 14]
```diff
@@ -107,42 +232,38 @@ class MinMaxByAggregate : public exec::Aggregate {
...
-      }
-      auto value = decodedValue.valueAt<T>(0);
-      auto comparisonValue = decodedComparisonValue.valueAt<U>(0);
-      auto nullValue = decodedValue.isNullAt(0);
-      rows.applyToSelected([&](vector_size_t i) {
-        updateValues(groups[i], value, comparisonValue, nullValue, mayUpdate);
-      });
-    } else if (
-        decodedValue.mayHaveNulls() || decodedComparisonValue.mayHaveNulls()) {
+    /// decodedValue will contain the values of column X.
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:235]: Use /// for method and class-level comments and // for in-the-body comments.

pre-existing, but would be nice to change "will contain" to "contains" and "which will be" -> "which is"

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 15]
```diff
@@ -203,95 +321,103 @@ class MinMaxByAggregate : public exec::Aggregate {
...
             false,
             mayUpdate);
       });
     }
   }
 
-  // Final aggregation will take (Value, comparisonValue) structs as inputs. It
-  // will produce the Value associated with the maximum/minimum of
-  // comparisonValue over all structs.
+  /// Final aggregation will take (Value, comparisonValue) structs as inputs. It
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:369]: nit: "will take" -> "takes", Value -> value

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 16]
```diff
@@ -203,95 +321,103 @@ class MinMaxByAggregate : public exec::Aggregate {
...
+  /// will produce the Value associated with the maximum/minimum of
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:370]: nit: "will produce" -> produces

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 17]
```diff
@@ -301,39 +427,52 @@ class MinMaxByAggregate : public exec::Aggregate {
...
   inline bool& valueIsNull(char* group) {
     return *reinterpret_cast<bool*>(
-        group + Aggregate::offset_ + sizeof(T) + sizeof(U));
+        group + Aggregate::offset_ + sizeof(ValueAccumulatorType) +
+        sizeof(ComparisonAccumulatorType));
   }
 
-  // Initial value will take the minimum and maximum values of the numerical
-  // limits.
+  /// Initial value will take the minimum and maximum values of the numerical
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:462]: nit: "will take" -> takes

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 18]
```diff
@@ -301,39 +427,52 @@ class MinMaxByAggregate : public exec::Aggregate {
...
+  /// limits. This is only meaningful if the comparison value is numeric type.
   const U initialValue_;
+  DecodedVector decodedValueVector_;
+  DecodedVector decodedComparisonVector_;
+  DecodedVector decodedIntermediateResultVector_;
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:467]: nit: you may want to drop Vector_ suffix for readability: decodedValue_, decodedComparison_, decodedIntermediateResult_

**Review by mbasmanova — COMMENTED** [review]
@xiaoxmeng The changes in the min_by/max_by aggregate function look good. These changes allow the function to take any type as 'value' including arrays, maps and structs. In a follow-up PR it would be nice to update the signature and add tests to confirm this use case works. 

I'll read the test next.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 19]
```diff
@@ -15,178 +15,1541 @@
  */
 #include <fmt/format.h>
 
+#include "velox/dwio/dwrf/test/utils/BatchMaker.h"
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:18]: BatchMaker is deprecated. Let's use VectorMaker instead. It is available via the base class and base class also has helper methods that makes it easier to use it.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 20]
```diff
@@ -15,178 +15,1541 @@
...
 #include "velox/exec/tests/utils/PlanBuilder.h"
 #include "velox/functions/prestosql/aggregates/AggregateNames.h"
 #include "velox/functions/prestosql/aggregates/tests/AggregationTestBase.h"
 
 using facebook::velox::exec::test::PlanBuilder;
+using namespace facebook::velox::exec::test;
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:24]: This makes the above line redundant.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 21]
```diff
@@ -15,178 +15,1541 @@
...
 
 namespace facebook::velox::aggregate::test {
 
 namespace {
 
+using facebook::velox::exec::test::MaterializedRow;
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:30]: Is this needed?

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 22]
```diff
@@ -15,178 +15,1541 @@
...
       const std::string& colName) {
-    std::string funcName = aggName == kMaxBy ? "max" : "min";
-
-    for (int i = 0; i < 6; ++i) {
-      auto valColName = fmt::format("c{}", i);
-
+    const std::string funcName = aggName == kMaxBy ? "max" : "min";
+    for (int i = 0; i < rowType_->size(); ++i) {
+      const auto valColName = fmt::format("c{}", i);
+      const std::string verifyDuckDbSql = "SELECT " + valColName +
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:41]: Looks like DuckDB now supports min_by and max_by aggregate functions. Let's simplify the test to use it. It would be helpful to submit a separate PR to update the existing test, then add more tests on top.

https://duckdb.org/docs/sql/aggregates

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 23]
```diff
@@ -15,178 +15,1541 @@
...
+          " FROM tmp WHERE " + colName + " = ( SELECT " + funcName + "(" +
+          colName + ") FROM tmp) LIMIT 1";
+      SCOPED_TRACE(fmt::format(
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:44]: I didn't know about SCOPED_TRACE. I read about it last night and it sounds very useful.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 24]
```diff
@@ -15,178 +15,1541 @@
...
+          "{}({}, {})\nverifyDuckDbSql: {}",
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:45]: this format duplicates the format call below; would be nice to refactor

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 25]
```diff
@@ -15,178 +15,1541 @@
...
-          "SELECT " + valColName + " FROM tmp WHERE " + colName +
-              " = ( SELECT " + funcName + "(" + colName +
-              ") FROM tmp) LIMIT 1");
+  void testGroupByAggregation(
+      const std::vector<RowVectorPtr>& vectors,
+      const std::string& aggName,
+      const std::string& colName) {
+    const std::string funcName = aggName == kMaxBy ? "max" : "min";
+    for (int volColIdx = 0; volColIdx < rowType_->size(); ++volColIdx) {
+      const int groupByColIdx = folly::Random::rand32(0, rowType_->size());
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:67]: Why random? It will make failures hard to reproduce? Why can't we just test all possible combinations?

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 26]
```diff
@@ -15,178 +15,1541 @@
...
+      const std::string groupByColName = rowType_->nameOf(groupByColIdx);
+      const std::string valColName = rowType_->nameOf(volColIdx);
+      const std::string verifyDuckDbSql = "SELECT " + groupByColName + ", " +
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:70]: This logic duplicates the format call below. Would be nice to refactor.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 27]
```diff
@@ -15,178 +15,1541 @@
...
+          aggName + "(" + valColName + ", " + colName + ") FROM tmp GROUP BY " +
+          groupByColName;
+      SCOPED_TRACE(fmt::format(
+          "{}({}, {}) GROUP BY {}\nverifyDuckDbSql: {}",
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:74]: ditto

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 28]
```diff
@@ -15,178 +15,1541 @@
...
-          op,
-          "SELECT " + valColName + " FROM tmp WHERE " + colName +
-              " = ( SELECT " + funcName + "(" + colName +
-              ") FROM tmp) LIMIT 1");
+  void testGroupByAggregation(
+      const std::vector<RowVectorPtr>& vectors,
+      const std::string& aggName,
+      const std::string& colName) {
+    const std::string funcName = aggName == kMaxBy ? "max" : "min";
+    for (int volColIdx = 0; volColIdx < rowType_->size(); ++volColIdx) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:66]: do not abbreviate: valueColumnIndex or valueIndex; same for other names

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 29]
```diff
@@ -15,178 +15,1541 @@
...
-TEST_F(MinMaxByAggregationTest, maxByPartialConst) {
-  // Const values.
-  auto vectors = {
-      makeRowVector({
-          makeConstant(5, 10),
-          makeConstant(10, 10),
-      }),
-  };
+TEST_F(MinMaxByAggregationTest, minMaxByPartialConst) {
+  for (bool isMin : {false, true}) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:105]: I wonder if it makes sense to use parameterized tests for this.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 30]
```diff
@@ -15,178 +15,1541 @@
...
+    struct {
+      const RowVectorPtr inputRowVector;
+      const std::string verifyDuckDbSql;
+      const bool expectedBuildFailure = false;
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:109]: Why build failure?

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 31]
```diff
@@ -15,178 +15,1541 @@
...
+          verifyDuckDbSql);
+    }
+  } testSettings[] = {
+    // Max by both numeric value and comparison types.
+    {makeRowVector(
+         {makeNullableFlatVector<int64_t>({1, 100, std::nullopt, std::nullopt}),
+          makeNullableFlatVector<int64_t>({20, 10, std::nullopt, 8})}),
+     "max_by(c0, c1)",
+     "SELECT struct_pack(x => 1, y => 20)"},
+#if 0
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:416]: Remove?

**Review by mbasmanova — COMMENTED** [review]
@xiaoxmeng Some comments for the tests. These are quite verbose and it is hard to tell what combinations are being testing. I wonder if there is a way to make this clearer.

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp (hunk 13 above)
Comment (xiaoxmeng) [reply velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:198]: rawNulls is defined for row vector at l183. I combine the two lines l197 and l198

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp (hunk 30 above)
Comment (xiaoxmeng) [reply velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:109]: It is verify that some data type we don't support it and expect an exception here when build velox operator plan.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp (hunk 29 above)
Comment (xiaoxmeng) [reply velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:105]: Sure,  good suggestion. Will do in followup per our offline discussion.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp (hunk 21 above)
Comment (xiaoxmeng) [reply velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:30]: Sorry it was there for debugging. Removed. Thanks!

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp (hunk 22 above)
Comment (xiaoxmeng) [reply velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:41]: Sure will do in a followup. Thanks!

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp (hunk 19 above)
Comment (xiaoxmeng) [reply velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:18]: Thanks and will make the change in the followup test refactor PR.

**Comment by mbasmanova** [comment]
@xiaoxmeng format-check CI job is failing. Would you take a look?

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp (hunk 19 above)
Comment (mbasmanova) [reply velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:18]: Actually, I don't see where this header is used. Is it? Would it be possible to first refactor the test in a separate PR? This would avoid a lot of churn in the code and make it easier to review.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 32]
```diff
@@ -179,6 +193,21 @@ class MinMaxByAggregationTestBase : public AggregationTestBase {
     return makeFlatVector<T>(numValues_, [](auto row) { return row + 1; });
   }
 
+  // Build a flat vector with StringView. The value in the returned flat vector
+  // is in ascending order.
+  template <>
+  FlatVectorPtr<StringView> buildDataVector() {
+    std::shared_ptr<std::string> value;
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:200]: no need to use shared_ptr, just regular std::string will work

**Review by mbasmanova — APPROVED** [review]
Nice. Thank you!
